packages/git/src/node/dugite-git.ts

```diff
@@ -0,0 +1,289 @@
+        }
+        return branch.name;
+    }
+
+    private getResetMode(mode: 'hard' | 'soft' | 'mixed') {
+        switch (mode) {
+            case 'hard': return GitResetMode.Hard;
+            case 'soft': return GitResetMode.Soft;
+            case 'mixed': return GitResetMode.Mixed;
+            default: throw new Error(`Unexpected Git reset mode: ${mode}.`);
+        }
+    }
+
+    private async mapStatus(toMap: IStatusResult, repository: Repository): Promise<WorkingDirectoryStatus> {
+        const repositoryPath = this.getFsPath(repository);
+        const aheadBehindPromise = this.mapAheadBehind(toMap.branchAheadBehind);
+        const changesPromise = this.mapFileChanges(toMap.workingDirectory, repositoryPath);
+        const aheadBehind = await aheadBehindPromise;
+        const changes = await changesPromise;
+        return {
+            exists: toMap.exists,
+            branch: toMap.currentBranch,
+            upstreamBranch: toMap.currentUpstreamBranch,
+            aheadBehind,
+            changes,
+            currentHead: toMap.currentTip
+        };
+    }
+
+    private async mapAheadBehind(toMap: IAheadBehind | undefined): Promise<{ ahead: number, behind: number } | undefined> {
+        return toMap ? { ...toMap } : undefined;
+    }
+
+    private async mapFileChanges(toMap: DugiteStatus, repositoryPath: string): Promise<GitFileChange[]> {
+        return Promise.all(toMap.files.map(file => this.mapFileChange(file, repositoryPath)));
+    }
+
+    private async mapFileChange(toMap: DugiteFileChange, repositoryPath: string): Promise<GitFileChange> {
+        const uriPromise = this.getUri(Path.join(repositoryPath, toMap.path));
+        const statusPromise = this.mapFileStatus(toMap.status);
+        const oldUriPromise = toMap.oldPath ? this.getUri(Path.join(repositoryPath, toMap.oldPath)) : undefined;
+        const uri = await uriPromise;
+        const status = await statusPromise;
+        const oldUri = await oldUriPromise;
+        return {
+            uri,
+            status,
+            oldUri,
+            staged: toMap.staged
+        };
+    }
+
+    private mapFileStatus(toMap: AppFileStatus): GitFileStatus {
+        switch (toMap) {
+            case AppFileStatus.Conflicted: return GitFileStatus.Conflicted;
+            case AppFileStatus.Copied: return GitFileStatus.Copied;
+            case AppFileStatus.Deleted: return GitFileStatus.Deleted;
+            case AppFileStatus.Modified: return GitFileStatus.Modified;
+            case AppFileStatus.New: return GitFileStatus.New;
+            case AppFileStatus.Renamed: return GitFileStatus.Renamed;
+            default: throw new Error(`Unexpected application file status: ${toMap}`);
+        }
+    }
+
+    private getFsPath(repository: Repository | string): string {
+        const uri = typeof repository === 'string' ? repository : repository.localUri;
+        return FileUri.fsPath(uri);
+    }
+
+    private getUri(path: string): string {
```

Comment:
this doesn't return a URI.. but is getUri... confusing a bit?

Reply:
It does return a uri as a string